README.md
# ChangeMapSettings

Allows you to change your map and map generation settings at any time you want. Note that map generation changes only apply to new chunks.

NOTE:
The 0.17 version of this mod is missing configuration options because those are missing from the lua api. Furthermore, frequency of water, trees and cliffs is inverted compared to the base game map generation screen. This is also an issue with the lua api.
The mod will remain in this state until the lua api is changed to properly expose everything.


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 5.0.2
Date: ????
  Changes:
---------------------------------------------------------------------------------------------------
Version: 5.0.1
Date: 2022-01-27
  Features:
    - Added peaceful mode description.
    - Changed license to MIT.
      I cannot (reasonably) play games without controller support.
      Maintaining mods for a game I can't play is an uninteresting hobby for me.
      So I want to make it easier for others to fork my mods.
      May you take this code and have fun with it.
  Bugfixes:
    - Fixed crash when setting map preset to "Space Exploration Defaults".
    - Fixed locale of labels for vanilla ores.
    - Fixed that space exploration planet size was reset to default when applying map gen settings. 
---------------------------------------------------------------------------------------------------
Version: 5.0.0
Date: 2021-02-11
  Features:
    - Updated to Factorio 1.1.
    - Added map generation preset selector.
    - Moved GUI into the center of the screen and made it draggable.
---------------------------------------------------------------------------------------------------
Version: 4.0.1
Date: 26. 05. 2020
  Minor Features:
    - Updated for 0.18.27.
---------------------------------------------------------------------------------------------------
Version: 4.0.0
Date: 25. 01. 2020
  Minor Features:
    - Updated for 0.18.
---------------------------------------------------------------------------------------------------
Version: 3.1.4
Date: 16. 07. 2019
  Minor Features:
    - Used new functions to restrict text fields to numbers only.
---------------------------------------------------------------------------------------------------
Version: 3.1.3
Date: 31. 05. 2019
  Bugfixes:
    - Fixed issue related to division by 0.
---------------------------------------------------------------------------------------------------
Version: 3.1.2
Date: 28. 05. 2019
  Changes:
    - Split map settings GUI into separate file.
---------------------------------------------------------------------------------------------------
Version: 3.1.1
Date: 24. 05. 2019
  Bugfixes:
    - Handled other mods nuking my GUI.
---------------------------------------------------------------------------------------------------
Version: 3.1.0
Date: 24. 05. 2019
  Features:
    - Finished the update to 0.17
---------------------------------------------------------------------------------------------------
Version: 3.0.5
Date: 30. 03. 2019
  Bugfixes:
    - Fixed that applying the map generation settings would set peaceful mode to false.
---------------------------------------------------------------------------------------------------
Version: 3.0.4
Date: 19. 03. 2019
  Bugfixes:
    - Fixed that the GUI could be cut off due to an undocumented internal limit.
---------------------------------------------------------------------------------------------------
Version: 3.0.3
Date: 15. 03. 2019
  Bugfixes:
    - Fixed that "Absorption modifier" and "Attack cost modifier" would not allow decimal values.
---------------------------------------------------------------------------------------------------
Version: 3.0.2
Date: 15. 03. 2019
  Minor Features:
    - Updated to reflect the base game changes in 0.17.12
---------------------------------------------------------------------------------------------------
Version: 3.0.1
Date: 13. 03. 2019
  Minor Features:
    - Thumbnail
---------------------------------------------------------------------------------------------------
Version: 3.0.0
Date: 13. 03. 2019
  Minor Features:
    - Update to 0.17
---------------------------------------------------------------------------------------------------
Version: 2.1.4
Date: 24. 02. 2019
  Bugfixes:
    - Fixed issue when cliff settings of the map were invalid
---------------------------------------------------------------------------------------------------
Version: 2.1.3
Date: 17. 10. 2018
  Bugfixes:
    - Fixed crash on factorissimo surfaces
---------------------------------------------------------------------------------------------------
Version: 2.1.2
Date: 28. 09. 2018
  Bugfixes:
    - Fixed handling of invalid autoplace controls in the map
---------------------------------------------------------------------------------------------------
Version: 2.1.1
Date: 26. 09. 2018
  Bugfixes:
    - Fixed issue with resources that were set to "none" frequency or richness
---------------------------------------------------------------------------------------------------
Version: 2.1.0
Date: 25. 09. 2018
  Major Features:
    - Added map generation options
---------------------------------------------------------------------------------------------------
Version: 2.0.0
Date: 11. 12. 2017
  Minor Features:
    - Update to 0.16
---------------------------------------------------------------------------------------------------
Version: 1.0.1
Date: 02. 10. 2017
  Minor Features:
    - Added peaceful mode option
---------------------------------------------------------------------------------------------------
Version: 1.0.0
Date: 29. 09. 2017
  Major Features:
    - Initial release.


gui.lua
local util = require("__ChangeMapSettings__/utilities")
local mod_gui = require("__core__/lualib/mod-gui")
local map_gen_gui = require("__ChangeMapSettings__/map_gen_settings_gui")
local map_settings_gui = require("__ChangeMapSettings__/map_settings_gui")
local gui = {}

-- GUI --
gui.regen = function(player)
  gui.kill(player)
  --toggle button
  local button_flow = mod_gui.get_button_flow(player)
  local button = button_flow.add{
    type = "button",
    name = "change-map-settings-toggle-config",
    style = mod_gui.button_style,
    tooltip = {"gui.change-map-settings-toggle-tooltip", {"gui.change-map-settings-title"}},
    caption = {"gui.change-map-settings-toggle-config-caption"}
  }
  button.visible = true

  -- general gui frame setup --
  local main_flow = player.gui.screen.add{
    type = "frame",
    name = "change-map-settings-main-flow",
    direction = "horizontal",
    style = "invisible_frame"
  }
  main_flow.visible = false
  main_flow.force_auto_center()

  -- map settings
  local map_settings_frame = main_flow.add{
    type = "frame",
    caption = {"gui.change-map-settings-title"},
    name = "change-map-settings-map-settings-frame",
    direction = "vertical"
  }
  map_settings_frame.drag_target = main_flow
  gui.make_map_settings(map_settings_frame, player.surface)

  -- start button at the bottom
  gui.make_start_button(map_settings_frame, false)


  -- map gen settings
  local map_gen_frame = main_flow.add{
    type = "frame",
    caption = {"gui.change-map-settings-map-gen-title"},
    name = "change-map-settings-map-gen-frame",
    direction = "vertical"
  }
  map_gen_frame.drag_target = main_flow
  gui.make_map_gen_settings(map_gen_frame)

  -- start button at the bottom
  gui.make_start_button(map_gen_frame, true)
end

gui.make_map_settings = function(parent, surface)
  local inner_frame = parent.add{
    type = "frame",
    direction = "vertical",
    style = "window_content_frame_packed"
  }

  -- tool buttons
  do
    local frame = inner_frame.add{
      type = "frame",
      direction = "horizontal",
      style = "subheader_frame"
    }
    frame.add{
      type = "flow",
      direction = "horizontal",
      style = "pusher"
    }
    frame.add{
      type = "sprite-button",
      name = "change-map-settings-use-current-button",
      style = "tool_button",
      sprite = "utility/refresh",
      tooltip = {"gui.change-map-settings-use-current-button-caption"}
    }
    frame.add{
      type = "sprite-button",
      name = "change-map-settings-default-button",
      style = "tool_button_red",
      sprite = "utility/reset",
      tooltip = {"gui.change-map-settings-default-button-caption"}
    }
  end

  local table_holder = inner_frame.add{
    type = "frame",
    name = "change-map-settings-map-settings-table-holder",
    style = "b_inner_frame_no_border"
  }

  local config_table = table_holder.add{
    type = "table",
    column_count = 2,
    vertical_centering = false
  }
  config_table.style.horizontal_spacing = 12

  local map_settings = game.map_settings
  --make different map gen option groups
  map_settings_gui.make_pollution_settings(config_table, map_settings)
  map_settings_gui.make_expansion_settings(config_table, map_settings)
  map_settings_gui.make_evolution_settings(config_table, map_settings)
  gui.make_general_map_settings(config_table, surface)
end

gui.make_general_map_settings = function(parent, surface)
  local config_more_option_general_flow = parent.add{
    type = "flow",
    name = "change-map-settings-config-more-general-flow",
    direction = "vertical"
  }
  config_more_option_general_flow.add{
    type = "label",
    caption = {"gui.change-map-settings-general-title"},
    style = "caption_label"
  }
  local config_more_option_general_table = config_more_option_general_flow.add{
    type = "table",
    name = "change-map-settings-config-more-general-table",
    column_count = 2,
    style = "bordered_table"
  }
  config_more_option_general_table.style.column_alignments[2] = "center"

  config_more_option_general_table.add{
    type = "label",
    caption = util.add_info_icon_to_string({"gui-map-generator.peaceful-mode-checkbox"}),
    tooltip = {"gui-map-generator.peaceful-mode-description"}
  }
  config_more_option_general_table.add{
    type = "checkbox",
    name = "change-map-settings-peaceful-checkbox",
    state = surface.peaceful_mode,
  }
  config_more_option_general_table.children[1].style.horizontally_stretchable = true
end

gui.get_map_settings_container = function(player)
  return player.gui.screen["change-map-settings-main-flow"]["change-map-settings-map-settings-frame"].children[1]["change-map-settings-map-settings-table-holder"].children[1]
end

gui.get_peaceful_mode_checkbox = function(player)
  return gui.get_map_settings_container(player)["change-map-settings-config-more-general-flow"]["change-map-settings-config-more-general-table"]["change-map-settings-peaceful-checkbox"]
end

gui.make_start_button = function(parent, map_gen)
  local start_button_flow = parent.add{
    type = "flow",
    direction = "horizontal"
  }
  start_button_flow.style.top_padding = 8
  start_button_flow.add{
    type = "flow",
    direction = "horizontal",
    style = "pusher"
  }
  start_button_flow.add{
    type = "button",
    name = "change-map-settings-start-" .. (map_gen and "map-gen-" or "") .. "button",
    tooltip = {"gui.change-map-settings-start-"  .. (map_gen and "map-gen-" or "") .. "button-tooltip"},
    caption = {"gui.change-map-settings-start-button-caption", {"gui.change-map-settings-"  .. (map_gen and "map-gen-" or "") .. "title"}},
    style = "confirm_button"
  }
end

gui.make_map_gen_settings = function(parent)
  local inner_frame = parent.add{
    type = "frame",
    direction = "vertical",
    style = "deep_frame"
  }

  do -- subheader
    local tool_button_frame = inner_frame.add{
      type = "frame",
      name = "change-map-settings-map-gen-button-frame",
      direction = "horizontal",
      style = "subheader_frame"
    }

    -- presets
    local presets = {}
    for name, preset in pairs(game.map_gen_presets) do
      if not preset.basic_settings then
        goto continue
      end
      if table_size(preset.basic_settings) == 1 and preset.basic_settings.property_expression_names and table_size(preset.basic_settings.property_expression_names) == 0 then
        goto continue -- vanilla marathon preset that does not have map gen settings affecting properties has an empty property_expression_names table for some reason
      end
      presets[#presets+1] = {"map-gen-preset-name." .. name}
      ::continue::
    end
    tool_button_frame.add{
      type = "drop-down",
      name = "change-map-settings-preset-dropdown",
      items = presets
    }
    tool_button_frame.add{
      type = "line",
      direction = "vertical"
    }

    -- seed
    local seed_label = tool_button_frame.add{
      type = "label",
      caption = {"gui.change-map-settings-seed-caption"},
      style = "caption_label"
    }
    seed_label.style.top_padding = 4
    seed_label.style.left_padding = 8
    tool_button_frame.add{
      type = "textfield",
      name = "change-map-settings-seed-textfield",
      text = "0",
      numeric = true,
      allow_decimal = false,
      allow_negative = false
    }
    tool_button_frame.add{
      type = "line",
      direction = "vertical"
    }

    -- tool_buttons
    tool_button_frame.add{
      type = "flow",
      direction = "horizontal",
      style = "pusher"
    }
    tool_button_frame.add{
      type = "sprite-button",
      name = "change-map-settings-use-current-map-gen-button",
      style = "tool_button",
      sprite = "utility/refresh",
      tooltip = {"gui.change-map-settings-use-current-button-caption"}
    }
    tool_button_frame.add{
      type = "sprite-button",
      name = "change-map-settings-default-map-gen-button",
      style = "tool_button_red",
      sprite = "utility/reset",
      tooltip = {"gui.change-map-settings-default-button-caption"}
    }
  end

  -- rest of map gen settings
  local map_gen_flow = inner_frame.add{
    type = "flow",
    name = "change-map-settings-map-gen-flow",
    direction = "horizontal"
  }
  map_gen_gui.create(map_gen_flow)
end

gui.get_map_gen_settings_container = function(player)
  return player.gui.screen["change-map-settings-main-flow"]["change-map-settings-map-gen-frame"].children[1]["change-map-settings-map-gen-flow"]
end

gui.get_seed_field = function(player)
  local tool_button_frame = player.gui.screen["change-map-settings-main-flow"]["change-map-settings-map-gen-frame"].children[1]["change-map-settings-map-gen-button-frame"]
  return tool_button_frame["change-map-settings-seed-textfield"]
end

gui.get_preset_dropdown = function(player)
  local tool_button_frame = player.gui.screen["change-map-settings-main-flow"]["change-map-settings-map-gen-frame"].children[1]["change-map-settings-map-gen-button-frame"]
  return tool_button_frame["change-map-settings-preset-dropdown"]
end

gui.kill = function(player)
  local button_flow = mod_gui.get_button_flow(player)
  if button_flow["change-map-settings-toggle-config"] then
    button_flow["change-map-settings-toggle-config"].destroy()
  end
  if player.gui.screen["change-map-settings-main-flow"] then
    player.gui.screen["change-map-settings-main-flow"].destroy()
  end

  --migration from pre 5.0.0
  local frame_flow = mod_gui.get_frame_flow(player)
  if frame_flow["change-map-settings-main-flow"] then
    frame_flow["change-map-settings-main-flow"].destroy()
  end

  --migration from 2.0.0
  if frame_flow["change-map-settings-config-more-frame"] then
		frame_flow["change-map-settings-config-more-frame"].destroy()
	end
end

return gui


data.lua
data.raw["gui-style"]["default"]["pusher"] =
{
  type = "horizontal_flow_style",
  horizontally_stretchable = "on"
}

data.raw["gui-style"]["default"]["deep_frame"] =
{
  type = "frame_style",
  parent = "inside_deep_frame",
  vertical_flow_style =
  {
    type = "vertical_flow_style",
    vertical_spacing = 8
  }
}

data.raw["gui-style"]["default"]["frame_in_deep_frame"] =
{
  type = "frame_style",
  parent = "frame",
  graphical_set =
  {
    base =
    {
      position = {51, 0}, corner_size = 8,
      center = {position = {76, 8}, size = {1, 1}},
      draw_type = "outer"
    },
    shadow = default_inner_shadow
  }
}

data.raw["gui-style"]["default"]["b_inner_frame_no_border"] =
{
  type = "frame_style",
  parent = "b_inner_frame",
  graphical_set =
  {
    base =
    {
      center = {position = {76, 8}, size = {1, 1}},
    },
    -- we only show shadow on the top, to solve the problem of it not being casted from the subheader panel above
    shadow =
    {
      top =
      {
        position = {191, 128},
        size = {1, 8},
        tint = hard_shadow_color,
        draw_type = "inner"
      }
    }
  }
}


map_settings_gui.lua
local util = require("__ChangeMapSettings__/utilities")
local MOD_PREFIX = "change-map-settings-"
local GUI_PREFIX = "map-settings-"
local ENTIRE_PREFIX = MOD_PREFIX .. GUI_PREFIX
local map_settings_gui = {}

map_settings_gui.make_pollution_settings = function(parent, map_settings)
  local WIDGET_PREFIX = "pollution-"
  local flow = parent.add{
    type = "flow",
    name = ENTIRE_PREFIX .. WIDGET_PREFIX .. "flow",
    direction = "vertical"
  }
  flow.add{
    type = "label",
    caption = {"gui-map-generator.pollution"},
    style = "caption_label"
  }
  local table = flow.add{
    type = "table",
    name = ENTIRE_PREFIX .. WIDGET_PREFIX .. "table",
    column_count = 2,
    style = "bordered_table"
  }
  table.style.column_alignments[2] = "center"

  table.add{
    type = "label",
    caption = {"gui-map-generator.pollution"}
  }
  table.add{
    type = "checkbox",
    name = ENTIRE_PREFIX .. WIDGET_PREFIX .. "checkbox",
    state = map_settings.pollution.enabled,
  }
  table.children[1].style.horizontally_stretchable = true
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "dissipation", {"gui-map-generator.pollution-absorption-modifier"}, {"gui-map-generator.pollution-absorption-modifier-description"}, tostring(map_settings.pollution.ageing), 50)
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "consumption", {"gui-map-generator.enemy-attack-pollution-consumption-modifier"}, {"gui-map-generator.enemy-attack-pollution-consumption-modifier-description"}, tostring(map_settings.pollution.enemy_attack_pollution_consumption_modifier), 50)
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "tree-dmg", {"gui-map-generator.minimum-pollution-to-damage-trees"}, {"gui-map-generator.minimum-pollution-to-damage-trees-description"}, tostring(map_settings.pollution.min_pollution_to_damage_trees), 50)
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "tree-absorb", {"gui-map-generator.pollution-absorbed-per-tree-damaged"}, {"gui-map-generator.pollution-absorbed-per-tree-damaged-description"}, tostring(map_settings.pollution.pollution_restored_per_tree_damage), 50)
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "diffusion", {"gui." .. MOD_PREFIX .. "in-unit", {"gui-map-generator.pollution-diffusion-ratio"}, {"gui." .. MOD_PREFIX .. "percent"}}, {"gui-map-generator.pollution-diffusion-ratio-description"}, tostring(map_settings.pollution.diffusion_ratio * 100), 50)
end

map_settings_gui.make_evolution_settings = function(parent, map_settings)
  local WIDGET_PREFIX = "evolution-"
  local flow = parent.add{
    type = "flow",
    name = ENTIRE_PREFIX .. WIDGET_PREFIX .. "flow",
    direction = "vertical"
  }
  flow.add{
    type = "label",
    caption = {"gui-map-generator.evolution"},
    style = "caption_label"
  }
  local table = flow.add{
    type = "table",
    name = ENTIRE_PREFIX .. WIDGET_PREFIX .. "table",
    column_count = 2,
    style = "bordered_table"
  }
  table.style.column_alignments[2] = "center"

  table.add{
    type = "label",
    caption = {"gui-map-generator.evolution"}
  }
  table.add{
    type = "checkbox",
    name = ENTIRE_PREFIX .. WIDGET_PREFIX .. "checkbox",
    state = map_settings.enemy_evolution.enabled,
  }
  table.children[1].style.horizontally_stretchable = true
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "factor", {"gui-map-generator.evolution"}, {"gui." .. MOD_PREFIX .. "evolution-factor-tooltip"}, util.number_to_string(game.forces["enemy"].evolution_factor), 80)
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "time", {"gui-map-generator.evolution-time-factor"}, {"gui-map-generator.evolution-time-factor-description"}, util.number_to_string(map_settings.enemy_evolution.time_factor * 100), 80)
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "destroy", {"gui-map-generator.evolution-destroy-factor"}, {"gui-map-generator.evolution-destroy-factor-description"}, util.number_to_string(map_settings.enemy_evolution.destroy_factor * 100), 80)
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "pollution", {"gui-map-generator.evolution-pollution-factor"}, {"gui-map-generator.evolution-pollution-factor-description"}, util.number_to_string(map_settings.enemy_evolution.pollution_factor * 100), 80)
end

map_settings_gui.make_expansion_settings = function(parent, map_settings)
  local WIDGET_PREFIX = "expansion-"
  local flow = parent.add{
    type = "flow",
    name = ENTIRE_PREFIX .. WIDGET_PREFIX .. "flow",
    direction = "vertical"
  }
  flow.add{
    type = "label",
    caption = {"gui-map-generator.enemy-expansion-group-tile"},
    style = "caption_label"
  }
  local table = flow.add{
    type = "table",
    name = ENTIRE_PREFIX .. WIDGET_PREFIX .. "table",
    column_count = 2,
    style = "bordered_table"
  }
  table.style.column_alignments[2] = "center"

  table.add{
    type = "label",
    caption = {"gui-map-generator.enemy-expansion-group-tile"}
  }
  table.add{
    type = "checkbox",
    name = ENTIRE_PREFIX .. WIDGET_PREFIX .. "checkbox",
    state = map_settings.enemy_expansion.enabled,
  }
  table.children[1].style.horizontally_stretchable = true
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "distance", {"gui-map-generator.enemy-expansion-maximum-expansion-distance"}, {"gui-map-generator.enemy-expansion-maximum-expansion-distance-description"}, tostring(map_settings.enemy_expansion.max_expansion_distance), 30)
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "min-size", {"gui-map-generator.enemy-expansion-minimum-expansion-group-size"}, {"gui-map-generator.enemy-expansion-minimum-expansion-group-size-description"}, tostring(map_settings.enemy_expansion.settler_group_min_size), 30)
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "max-size", {"gui-map-generator.enemy-expansion-maximum-expansion-group-size"}, {"gui-map-generator.enemy-expansion-maximum-expansion-group-size-description"}, tostring(map_settings.enemy_expansion.settler_group_max_size), 30)
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "min-cd", {"gui." .. MOD_PREFIX .. "in-unit", {"gui-map-generator.enemy-expansion-minimum-expansion-cooldown"}, {"gui." .. MOD_PREFIX .. "minutes"}}, {"gui-map-generator.enemy-expansion-minimum-expansion-cooldown-description"},  tostring(map_settings.enemy_expansion.min_expansion_cooldown / 3600), 30)
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "max-cd", {"gui." .. MOD_PREFIX .. "in-unit", {"gui-map-generator.enemy-expansion-maximum-expansion-cooldown"}, {"gui." .. MOD_PREFIX .. "minutes"}}, {"gui-map-generator.enemy-expansion-maximum-expansion-cooldown-description"}, tostring(map_settings.enemy_expansion.max_expansion_cooldown / 3600), 30)
end

map_settings_gui.make_config_option = function(parent, name, caption, tooltip, default, max_width)
  parent.add{
    type = "label",
    caption = util.add_info_icon_to_string(caption),
    tooltip = tooltip
  }
  local child = parent.add{
    type = "textfield",
    name = ENTIRE_PREFIX .. name .. "-textfield",
    text = default,
    numeric = true,
    allow_decimal = true,
    allow_negative = false
  }
  if max_width then child.style.maximal_width = max_width end
  return child
end

map_settings_gui.expansion_reset_to_defaults = function(parent)
  local WIDGET_PREFIX = "expansion-"
  local table = parent[ENTIRE_PREFIX .. WIDGET_PREFIX .. "flow"][ENTIRE_PREFIX .. WIDGET_PREFIX .. "table"]
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "checkbox"].state = true
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "distance-textfield"].text = "7"
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "min-size-textfield"].text = "5"
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "max-size-textfield"].text = "20"
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "min-cd-textfield"].text = "4"
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "max-cd-textfield"].text = "60"
end

map_settings_gui.evolution_reset_to_defaults = function(parent)
  local WIDGET_PREFIX = "evolution-"
  local table = parent[ENTIRE_PREFIX .. WIDGET_PREFIX .. "flow"][ENTIRE_PREFIX .. WIDGET_PREFIX .. "table"]
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "checkbox"].state = true
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "factor-textfield"].text = "0"
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "time-textfield"].text = "0.0004"
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "destroy-textfield"].text = "0.2"
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "pollution-textfield"].text = "0.000090"
end

map_settings_gui.pollution_reset_to_defaults = function(parent)
  local WIDGET_PREFIX = "pollution-"
  local table = parent[ENTIRE_PREFIX .. WIDGET_PREFIX .. "flow"][ENTIRE_PREFIX .. WIDGET_PREFIX .. "table"]
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "checkbox"].state = true
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "dissipation-textfield"].text = "1"
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "consumption-textfield"].text = "1"
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "tree-dmg-textfield"].text = "60"
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "tree-absorb-textfield"].text = "10"
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "diffusion-textfield"].text = "2"
end

map_settings_gui.expansion_set_to_current = function(parent, map_settings)
  local WIDGET_PREFIX = "expansion-"
  local table = parent[ENTIRE_PREFIX .. WIDGET_PREFIX .. "flow"][ENTIRE_PREFIX .. WIDGET_PREFIX .. "table"]
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "checkbox"].state = map_settings.enemy_expansion.enabled
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "distance-textfield"].text = tostring(map_settings.enemy_expansion.max_expansion_distance)
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "min-size-textfield"].text = tostring(map_settings.enemy_expansion.settler_group_min_size)
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "max-size-textfield"].text = tostring(map_settings.enemy_expansion.settler_group_max_size)
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "min-cd-textfield"].text = tostring(map_settings.enemy_expansion.min_expansion_cooldown / 3600)
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "max-cd-textfield"].text = tostring(map_settings.enemy_expansion.max_expansion_cooldown / 3600)
end

map_settings_gui.evolution_set_to_current = function(parent, map_settings)
  local WIDGET_PREFIX = "evolution-"
  local table = parent[ENTIRE_PREFIX .. WIDGET_PREFIX .. "flow"][ENTIRE_PREFIX .. WIDGET_PREFIX .. "table"]
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "checkbox"].state = map_settings.enemy_evolution.enabled
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "factor-textfield"].text = util.number_to_string(game.forces["enemy"].evolution_factor)
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "time-textfield"].text = util.number_to_string(map_settings.enemy_evolution.time_factor * 100)
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "destroy-textfield"].text = util.number_to_string(map_settings.enemy_evolution.destroy_factor * 100)
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "pollution-textfield"].text = util.number_to_string(map_settings.enemy_evolution.pollution_factor * 100)
end

map_settings_gui.pollution_set_to_current = function(parent, map_settings)
  local WIDGET_PREFIX = "pollution-"
  local table = parent[ENTIRE_PREFIX .. WIDGET_PREFIX .. "flow"][ENTIRE_PREFIX .. WIDGET_PREFIX .. "table"]
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "checkbox"].state = map_settings.pollution.enabled
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "dissipation-textfield"].text = tostring(map_settings.pollution.ageing)
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "consumption-textfield"].text = tostring(map_settings.pollution.enemy_attack_pollution_consumption_modifier)
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "tree-dmg-textfield"].text = tostring(map_settings.pollution.min_pollution_to_damage_trees)
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "tree-absorb-textfield"].text = tostring(map_settings.pollution.pollution_restored_per_tree_damage)
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "diffusion-textfield"].text = tostring(map_settings.pollution.diffusion_ratio * 100)
end

-- can throw!
map_settings_gui.expansion_read = function(parent)
  local WIDGET_PREFIX = "expansion-"
  local table = parent[ENTIRE_PREFIX .. WIDGET_PREFIX .. "flow"][ENTIRE_PREFIX .. WIDGET_PREFIX .. "table"]
  local enemy_expansion = {}

  enemy_expansion.enabled = table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "checkbox"].state
  enemy_expansion.max_expansion_distance = util.check_bounds(util.textfield_to_uint(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "distance-textfield"]),
                                                             2, 20,
                                                             {"msg." .. MOD_PREFIX .. "invalid-expansion-distance"})
  enemy_expansion.settler_group_min_size = util.check_bounds(util.textfield_to_uint(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "min-size-textfield"]),
                                                             1, 20,
                                                             {"msg." .. MOD_PREFIX .. "invalid-expansion-min-size"})
  enemy_expansion.settler_group_max_size = util.check_bounds(util.textfield_to_uint(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "max-size-textfield"]),
                                                             math.max(enemy_expansion.settler_group_min_size, 1), 50,
                                                             {"msg." .. MOD_PREFIX .. "invalid-expansion-max-size"})
  enemy_expansion.min_expansion_cooldown = util.check_bounds(util.textfield_to_uint(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "min-cd-textfield"]),
                                                             1, 60,
                                                             {"msg." .. MOD_PREFIX .. "invalid-expansion-min-cd"}) * 3600
  enemy_expansion.max_expansion_cooldown = util.check_bounds(util.textfield_to_uint(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "max-cd-textfield"]),
                                                             math.max(enemy_expansion.min_expansion_cooldown / 3600, 5), 180,
                                                             {"msg." .. MOD_PREFIX .. "invalid-expansion-max-cd"}) * 3600
  return enemy_expansion
end

-- can throw!
map_settings_gui.evolution_read = function(parent)
  local WIDGET_PREFIX = "evolution-"
  local table = parent[ENTIRE_PREFIX .. WIDGET_PREFIX .. "flow"][ENTIRE_PREFIX .. WIDGET_PREFIX .. "table"]
  local enemy_evolution = {}

  enemy_evolution.enabled = table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "checkbox"].state
  enemy_evolution.evolution_factor = util.check_bounds(util.textfield_to_number(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "factor-textfield"]),
                                                            0, 1,
                                                            {"msg." .. MOD_PREFIX .. "invalid-evolution-factor"})
  enemy_evolution.time_factor = util.check_bounds(util.textfield_to_number(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "time-textfield"]),
                                                  0, 0.01,
                                                  {"msg." .. MOD_PREFIX .. "invalid-evolution-time"}) / 100
  enemy_evolution.destroy_factor = util.check_bounds(util.textfield_to_number(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "destroy-textfield"]),
                                                     0, 1,
                                                     {"msg." .. MOD_PREFIX .. "invalid-evolution-destroy"}) / 100
  enemy_evolution.pollution_factor = util.check_bounds(util.textfield_to_number(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "pollution-textfield"]),
                                                       0, 0.01,
                                                       {"msg." .. MOD_PREFIX .. "invalid-evolution-pollution"}) / 100
  return enemy_evolution
end

-- can throw!
map_settings_gui.pollution_read = function(parent)
  local WIDGET_PREFIX = "pollution-"
  local table = parent[ENTIRE_PREFIX .. WIDGET_PREFIX .. "flow"][ENTIRE_PREFIX .. WIDGET_PREFIX .. "table"]
  local pollution = {}

  pollution.enabled = table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "checkbox"].state
  pollution.ageing = util.check_bounds(util.textfield_to_number(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "dissipation-textfield"]),
                                       0.1, 4,
                                       {"msg." .. MOD_PREFIX .. "invalid-pollution-absorption"})
  pollution.enemy_attack_pollution_consumption_modifier = util.check_bounds(util.textfield_to_number(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "consumption-textfield"]),
                                                                            0.1, 4,
                                                                            {"msg." .. MOD_PREFIX .. "invalid-enemy-attack-pollution-consumption"})
  pollution.min_pollution_to_damage_trees = util.check_bounds(util.textfield_to_uint(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "tree-dmg-textfield"]),
                                                              0, 9999,
                                                              {"msg." .. MOD_PREFIX .. "invalid-pollution-tree-dmg"})
  pollution.pollution_restored_per_tree_damage = util.check_bounds(util.textfield_to_uint(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "tree-absorb-textfield"]),
                                                                   0, 9999,
                                                                   {"msg." .. MOD_PREFIX .. "invalid-pollution-tree-absorb"})
  pollution.diffusion_ratio = util.check_bounds(util.textfield_to_uint(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "diffusion-textfield"]),
                                                0, 25,
                                                {"msg." .. MOD_PREFIX .. "invalid-pollution-diffusion"}) / 100
  return pollution
end

return map_settings_gui
